[Feature] Adds utils method remove_duplicates

albertbou92 opened the pull request:
## Description

 The method detects duplicate elements in the tensor associated with the specified 'key' along the specified
'dim' and removes elements in the same indices in all other tensors within the TensorDict.

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/tensordict/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/pytorch/tensordict/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

vmoens commented on tensordict/utils.py:
```diff
@@ -1862,3 +1862,65 @@ def _index_preserve_data_ptr(index):
     return False
+
+
+def remove_duplicates(
```
to be added in the doc

vmoens commented on tensordict/utils.py:
```diff
@@ -1862,3 +1862,65 @@ def _index_preserve_data_ptr(index):
+) -> TensorDictBase:
+    """Removes indices duplicated in 'key' along the specified dimension.
+
+    This method detects duplicate elements in the tensor associated with the specified 'key' along the specified
+    'dim' and removes elements in the same indices in all other tensors within the TensorDict.
```
```suggestion
    This method detects duplicate elements in the tensor associated with the specified `key` along the specified
    `dim` and removes elements in the same indices in all other tensors within the TensorDict.
```

vmoens replied:
We should make it clear that this all happens along the batch-size (ie, empty batch size will throw an error etc)

vmoens commented on tensordict/utils.py:
```diff
@@ -1862,3 +1862,65 @@ def _index_preserve_data_ptr(index):
+    Args:
+        tensordict (TensorDictBase): The TensorDict containing potentially duplicate elements.
+        key (str): The key of the tensor along which duplicate elements should be identified and removed.
+        dim (int): The dimension along which duplicate elements should be identified and removed. Default: 0.
```
```suggestion
        dim (int): The dimension along which duplicate elements should be identified and removed.
            Defaults to ``0``.
```
Also I guess dim must be within the batch-size right?

vmoens commented on tensordict/utils.py:
```diff
@@ -1862,3 +1862,65 @@ def _index_preserve_data_ptr(index):
+        dim (int): The dimension along which duplicate elements should be identified and removed. Default: 0.
+
+    Returns:
+        TensorDictBase: tensordict with removed elements.
```
tell more about the batch size of the resulting td

vmoens commented on tensordict/utils.py:
```diff
@@ -1862,3 +1862,65 @@ def _index_preserve_data_ptr(index):
+        TensorDictBase: tensordict with removed elements.
+
+    Example:
+        >>> tensordict = TensorDict(
+        ...     {
+        ...         "tensor1": torch.tensor([[1, 2, 3], [4, 5, 6], [1, 2, 3], [7, 8, 9]]),
+        ...         "tensor2": torch.tensor([[10, 20], [30, 40], [40, 50], [50, 60]]),
+        ...     }
+        ... )
```
this one has no batch-size

vmoens commented on tensordict/utils.py:
```diff
@@ -1862,3 +1862,65 @@ def _index_preserve_data_ptr(index):
+        ...     },
+        ...     batch_size=[3],
+        ... )
+        >>> for key in output_tensordict.keys():
+        ...     assert torch.equal(output_tensordict[key], expected_output[key])
```
```suggestion
        >>> assert (tensordict == expected_output).all()
```

vmoens commented on tensordict/utils.py:
```diff
@@ -1862,3 +1862,65 @@ def _index_preserve_data_ptr(index):
+        raise KeyError(f"The key '{key}' does not exist in the TensorDict.")
+
+    # Check dim is valid
+    if dim >= len(tensordict.batch_size):
```
we should make it work for neg dims, like -1 (last dim of the batch size). Check other methods like unbind etc to see how to deal with that

vmoens commented on tensordict/utils.py:
```diff
@@ -1862,3 +1862,65 @@ def _index_preserve_data_ptr(index):
+
+    Args:
+        tensordict (TensorDictBase): The TensorDict containing potentially duplicate elements.
+        key (str): The key of the tensor along which duplicate elements should be identified and removed.
```
We should make it extra clear that key must point to a tensor, not a tensordict (ie it must be a leaf)

vmoens replied:
```suggestion
        key (NestedKey): The key of the tensor along which duplicate elements should be identified and removed.
```

vmoens commented on tensordict/utils.py:
```diff
@@ -1862,3 +1862,65 @@ def _index_preserve_data_ptr(index):
+    first_indices = unique_indices_sorted[cum_sum]
+
+    # Remove duplicate elements in the TensorDict
+    unique_tensordict = tensordict[first_indices]
```
can we get the indices in the output tensordict (optionally)? like `return_indices=False` by default. The key could be `"unique_indices"`, maybe parametrizable (up to you, not sure if that makes sense). For sure we don't want to override an existing key so whatever key we choose we must check it isn't there already.

albertbou92 replied:
We can but since the indices are an 1D tensor of length batch_size[dim], it is not guaranteed that we can add them to the tensordict. We can return them separately (optionally), which also solves the problem of potentially override an existing key.

vmoens commented on test/test_utils.py:
```diff
@@ -151,6 +152,43 @@ def test_unravel_key_to_tuple():
 
 
+@pytest.mark.parametrize("key", ("tensor1", "tensor3"))
+@pytest.mark.parametrize("dim", (0, 1))
```
```suggestion
@pytest.mark.parametrize("dim", (0, 1, -1, -2))
```

vmoens commented on test/test_utils.py:
```diff
@@ -151,6 +152,43 @@ def test_unravel_key_to_tuple():
 
+@pytest.mark.parametrize("key", ("tensor1", "tensor3"))
+@pytest.mark.parametrize("dim", (0, 1))
+def test_remove_duplicates(key, dim):
```
I would create a td with nested keys too to make sure everything works ok

vmoens reviewed (commented):
Great work! I left some suggestions to start the discussion

albertbou92 commented:
I added the option to return the indices. 
I also handled the case of negative dims, which are now accepted.
finally I added some tests for the case of TDs of more than 1 dimension.

vmoens commented on tensordict/utils.py:
```diff
@@ -1862,3 +1862,89 @@ def _index_preserve_data_ptr(index):
+        tensordict (TensorDictBase): The TensorDict containing potentially duplicate elements.
+        key (NestedKey): The key of the tensor along which duplicate elements should be identified and removed. It
+            must be one of the leaf keys within the TensorDict, pointing to a tensor and not to another TensorDict.
+        dim (int): The dimension along which duplicate elements should be identified and removed. It must be one of
```
Int, optional

vmoens commented on tensordict/utils.py:
```diff
@@ -1862,3 +1862,89 @@ def _index_preserve_data_ptr(index):
+        raise KeyError(f"The key '{key}' does not point to a tensor in the TensorDict.")
+
+    # Check dim is valid
+    batch_dims = len(tensordict.batch_size)
```
Or tendordict.ndim

albertbou92 commented:
I added tests on different devices, turned return_indices into a kwarg only and fixed the other minor comments.

Regarding adding `unique_indices` to the output tensordict, `unique_indices` will also be a tensor of size `input.batch_size[dim]`. There is no guarantee that we can add this to the output TensorDict due to the shape (only if dim=0 and len(batch_size) = 1). This is why I decided to return it as an extra output.

vmoens commented:
> I added tests on different devices, turned return_indices into a kwarg only and fixed the other minor comments.
> 
> Regarding adding `unique_indices` to the output tensordict, `unique_indices` will also be a tensor of size `input.batch_size[dim]`. There is no guarantee that we can add this to the output TensorDict due to the shape (only if dim=0 and len(batch_size) = 1). This is why I decided to return it as an extra output. is there any alternative?

Good point. The alternative would be to expand it on the left but it's messy. Happy with the PR in its current state then!

vmoens commented on tensordict/utils.py:
```diff
@@ -1862,3 +1862,90 @@ def _index_preserve_data_ptr(index):
+    # Find first occurrence of each index  (e.g. [0, 1, 3])
+    _, unique_indices_sorted = torch.sort(unique_indices, stable=True)
+    cum_sum = counts.cumsum(0)
+    cum_sum = torch.cat((torch.tensor([0], device=input.device), cum_sum[:-1]))
```
```suggestion
    cum_sum = torch.cat((torch.zeros(1, device=input.device), cum_sum[:-1]))
```

vmoens commented on tensordict/utils.py:
```diff
@@ -1862,3 +1862,90 @@ def _index_preserve_data_ptr(index):
+        ... )
+        >>> assert (td == expected_output).all()
+    """
+    # Check if the key is a TensorDict
+    if key not in input.keys(include_nested=True):
+        raise KeyError(f"The key '{key}' does not exist in the TensorDict.")
+
+    # Check that the key points to a tensor
+    if not isinstance(input.get(key), torch.Tensor):
+        raise KeyError(f"The key '{key}' does not point to a tensor in the TensorDict.")
```
```suggestion
    tensor = input.get(key, default=None)

    # Check if the key is a TensorDict
    if tensor is None:
        raise KeyError(f"The key '{key}' does not exist in the TensorDict.")

    # Check that the key points to a tensor
    if not isinstance(tensor, torch.Tensor):
        raise KeyError(f"The key '{key}' does not point to a tensor in the TensorDict.")
```

vmoens commented on tensordict/utils.py:
```diff
@@ -1862,3 +1862,90 @@ def _index_preserve_data_ptr(index):
+        dim = batch_dims + dim
+
+    # Get indices of unique elements (e.g. [0, 1, 0, 2])
+    tensor = input.get(key)
```
```suggestion
```

vmoens reviewed (commented):
some small suggestions
